File: v3/pkg/application/http.go
package application

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"
)

// HTTPRequest represents an HTTP request from the frontend
type HTTPRequest struct {
	Method  string            `json:"method"`
	URL     string            `json:"url"`
	Headers map[string]string `json:"headers,omitempty"`
	Body    string            `json:"body,omitempty"`
	Timeout int               `json:"timeout,omitempty"` // timeout in seconds
}

// HTTPResponse represents an HTTP response to send back to the frontend
type HTTPResponse struct {
	StatusCode int               `json:"statusCode"`
	Headers    map[string]string `json:"headers"`
	Body       string            `json:"body"`
	Error      string            `json:"error,omitempty"`
}

// PerformHTTPRequest performs an HTTP request on behalf of the frontend
func PerformHTTPRequest(request HTTPRequest) HTTPResponse {
	// Validate method
	method := strings.ToUpper(request.Method)
	if method != "GET" && method != "POST" && method != "PUT" && method != "DELETE" && method != "PATCH" && method != "HEAD" && method != "OPTIONS" {
		return HTTPResponse{
			StatusCode: 0,
			Error:      fmt.Sprintf("Invalid HTTP method: %s", request.Method),
		}
	}

	// Create HTTP client with timeout
	timeout := 30 * time.Second
	if request.Timeout > 0 {
		timeout = time.Duration(request.Timeout) * time.Second
	}
	client := &http.Client{
		Timeout: timeout,
	}

	// Create request
	var body io.Reader
	if request.Body != "" {
		body = bytes.NewBufferString(request.Body)
	}

	req, err := http.NewRequestWithContext(context.Background(), method, request.URL, body)
	if err != nil {
		return HTTPResponse{
			StatusCode: 0,
			Error:      fmt.Sprintf("Failed to create request: %v", err),
		}
	}

	// Set headers
	for key, value := range request.Headers {
		req.Header.Set(key, value)
	}

	// Set default User-Agent if not provided
	if req.Header.Get("User-Agent") == "" {
		req.Header.Set("User-Agent", "Wails/3.0")
	}

	// Perform the request
	resp, err := client.Do(req)
	if err != nil {
		return HTTPResponse{
			StatusCode: 0,
			Error:      fmt.Sprintf("Request failed: %v", err),
		}
	}
	defer resp.Body.Close()

	// Read response body
	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return HTTPResponse{
			StatusCode: resp.StatusCode,
			Error:      fmt.Sprintf("Failed to read response body: %v", err),
		}
	}

	// Build response headers
	responseHeaders := make(map[string]string)
	for key, values := range resp.Header {
		if len(values) > 0 {
			responseHeaders[key] = values[0]
		}
	}

	return HTTPResponse{
		StatusCode: resp.StatusCode,
		Headers:    responseHeaders,
		Body:       string(bodyBytes),
	}
}

// parseHTTPRequest parses the HTTP request from JSON
func parseHTTPRequest(data string) (*HTTPRequest, error) {
	var request HTTPRequest
	err := json.Unmarshal([]byte(data), &request)
	if err != nil {
		return nil, fmt.Errorf("failed to parse HTTP request: %v", err)
	}

	// Validate required fields
	if request.URL == "" {
		return nil, fmt.Errorf("URL is required")
	}
	if request.Method == "" {
		request.Method = "GET"
	}

	return &request, nil
}

File: v3/internal/runtime/desktop/http.js
/*
 _	   __	  _ __
| |	 / /___ _(_) /____
| | /| / / __ `/ / / ___/
| |/ |/ / /_/ / / (__  )
|__/|__/\__,_/_/_/____/
The electron alternative for Go
(c) Lea Anthony 2019-present
*/

/* jshint esversion: 9 */

import {newRuntimeCallerWithID, objectNames} from "./runtime";

let call = newRuntimeCallerWithID(objectNames.HTTP);

let HTTPFetch = 0;

/**
 * Perform an HTTP request
 * @param {Object} options - The request options
 * @param {string} options.url - The URL to request
 * @param {string} [options.method='GET'] - The HTTP method
 * @param {Object} [options.headers] - Request headers
 * @param {string} [options.body] - Request body
 * @param {number} [options.timeout] - Request timeout in seconds
 * @returns {Promise<Object>} The response object
 */
export function Fetch(options) {
    // Ensure we have required fields
    if (!options || !options.url) {
        return Promise.reject(new Error("URL is required"));
    }

    // Set defaults
    const request = {
        url: options.url,
        method: options.method || 'GET',
        headers: options.headers || {},
        body: options.body || '',
        timeout: options.timeout || 30
    };

    // For POST requests, we need to send the body differently
    if (request.body) {
        return call(HTTPFetch, null, JSON.stringify(request));
    }

    return call(HTTPFetch, null, JSON.stringify(request));
}

/**
 * Convenience method for GET requests
 * @param {string} url - The URL to request
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Get(url, options = {}) {
    return Fetch({ ...options, url, method: 'GET' });
}

/**
 * Convenience method for POST requests
 * @param {string} url - The URL to request
 * @param {string|Object} body - The request body
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Post(url, body, options = {}) {
    if (typeof body === 'object' && !(body instanceof String)) {
        body = JSON.stringify(body);
        options.headers = {
            'Content-Type': 'application/json',
            ...(options.headers || {})
        };
    }
    return Fetch({ ...options, url, method: 'POST', body });
}

/**
 * Convenience method for PUT requests
 * @param {string} url - The URL to request
 * @param {string|Object} body - The request body
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Put(url, body, options = {}) {
    if (typeof body === 'object' && !(body instanceof String)) {
        body = JSON.stringify(body);
        options.headers = {
            'Content-Type': 'application/json',
            ...(options.headers || {})
        };
    }
    return Fetch({ ...options, url, method: 'PUT', body });
}

/**
 * Convenience method for DELETE requests
 * @param {string} url - The URL to request
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Delete(url, options = {}) {
    return Fetch({ ...options, url, method: 'DELETE' });
}

/**
 * Convenience method for PATCH requests
 * @param {string} url - The URL to request
 * @param {string|Object} body - The request body
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Patch(url, body, options = {}) {
    if (typeof body === 'object' && !(body instanceof String)) {
        body = JSON.stringify(body);
        options.headers = {
            'Content-Type': 'application/json',
            ...(options.headers || {})
        };
    }
    return Fetch({ ...options, url, method: 'PATCH', body });
}

/**
 * Convenience method for HEAD requests
 * @param {string} url - The URL to request
 * @param {Object} [options] - Additional options
 * @returns {Promise<Object>} The response object
 */
export function Head(url, options = {}) {
    return Fetch({ ...options, url, method: 'HEAD' });
}